Basic Legion & Hivelord

Jacquerel opened the pull request:
## About The Pull Request

Converts Hivelords and Legions to basic mobs (they were the same typepath) and also separates their behaviour.

![dreamseeker_kUK5linLkh](https://github.com/tgstation/tgstation/assets/7483112/72d27df7-22b9-4868-b101-061d5ce0049a)
**Hivelords** act much like before and spawn little worms next to them (with a new animation) which chase after you.
Their _new_ behaviour is that they also spawn worms whenever they take damage, and even more when they die.
Because their mobs are significantly less dangerous than those of a Legion, being swarmed with them is less immediately a death sentence.

![dreamseeker_Yb0GCN2dQp](https://github.com/tgstation/tgstation/assets/7483112/f5edb4a3-399d-434e-b58a-1b60f51a8e69)
**Legions** on the other hand have learned to throw their brood across long distances and are now a sort of spawner artillery.
This has advantages and disadvantages for them, the skulls spawn _closer_ to you but it also means that they're not body-blocking for the legion nearly as often.
Additionally, legions will prioritise firing skulls at their _allies_ in order to heal them (although they cannot heal other legions).
If for instance you were doing some kind of wave-based defence minigame then a Legion would run to the sidelines and then act as a ranged support until you chased it down.

Being eaten by a Legion has also had a few changes.
Consumed mobs are actually fulled healed and put into stasis as legions are inherently regenerative, so if your corpse is retrieved it's actually much faster to revive you because your corpse won't be damaged. Inside a Legion you won't decay or be ravaged by storms. Additionally, the Legion will attempt to lure your allies to Lavaland by occasionally speaking over the radio, repeating something that you said before your death.

It's not all good news though, because some of the time they'll leave you with an underlying condition which, if left untreated after your revival, will have you eventually turn back into a Legion again.

In order to support this I additionally modified the behaviour of the Crew Monitoring Console, because they way it worked was sort of unintuitive.
Previously it would estimate your current life status as a human with suit sensors maximised based entirely upon damage taken rather than if you were dead or not. Now if you are dead but unharmed, it will show a _green_ skull icon next to you instead of a red one.

Additional code changes I made throughout this PR:
- Mob fleeing now uses a blackboard key rather than a var on the behaviour to decide how far they should run. We should really use blackboard keys over vars when possible, it's silly to make behaviour subtypes when we have an existing solution for mob-specific data.
- AI targetting datums can optionally read an "aggro range" blackboard key. It won't select _new_ targets which are further than its aggro range, however it can maintain existing targets up to the (presumably bigger) existing vision value (should probably also be a blackboard key...)
- Added a couple of generic controllers for very basic AI which was shared between some mobs we already have.
- The "appearance_on_aggro" component can also change an icon state.
- Moved the code duplicated between changeling absorb and mindread to copy a player's last few messages into a generic mob helper, so that it wouldn't be duplicated in _three_ places.
- Reintroduced darkvision to basic mining mobs, it vanished at some point (if it was ever there).
- Player controlled goliath has click controls that I forgot to give it.

I also noticed that the way we do random speech is... not great. I'm going to refactor those behaviours entirely in a separate PR I think. This one is already big enough.

## Why It's Good For The Game

![image](https://github.com/tgstation/tgstation/assets/7483112/fbc16a65-b269-4181-ada5-e0e80c8ff82c)
Wahoo we did it that's all* of the lavaland mining mobs
<sup><sub>*forget about the elites and megafauna</sub></sup>

The changes are _mostly_ just things I think are cool and/or soulful
Legions are slightly more aggressive, but in this form I think will actually be _less_ annoying for ArcMining wave defence and also commit less time to defending themselves (aside from running away)
Having consumed miners murmur over the radio is both cool and also makes it marginally more likely someone will realise that they have died.

## Changelog

:cl:
refactor: Hivelords and Legions now use the basic mob framework. Please report any unusual behaviour.
add: Hivelords shed more spawn when they are attacked.
add: Legions have learned how to fling their skulls across long distances.
add: Legions can heal other lavaland mobs with their skulls.
add: Legions are better at preserving corpses they consume, and sometimes make use of their radios.
add: Legions may leave behind an unpleasant surprise after you are rescued from them.
balance: The crew monitoring console will now display you as dead if you are dead, an critically injured if in crit, rather than setting those icons purely based on your current health.
qol: You won't continue burning to a husk if consumed by a snow legion after being set on fire by an ice drake.
/:cl:

Jacquerel committed:
Hivelord

Jacquerel committed:
Merge branch 'master' into im-legion-no-im-legion

Jacquerel committed:
Crew monitor always shows a skull if you are dead even if you have full HP
Legion mob with no AI or abilities

Jacquerel committed:
Funny medical surprise

Jacquerel committed:
launch ability

Jacquerel committed:
AI & removal

Jacquerel committed:
Legions can use the radio

Jacquerel committed:
tweaks to speech copy
unify legion corpse spawners

Jacquerel committed:
Bug fixes and tuning and such
lol i forgot to give legions hp

Jacquerel committed:
Merge branch 'master' into im-legion-no-im-legion

Jacquerel committed:
How long has lavaland mob darkvision been missing?

Jacquerel committed:
Tweaks to legion tumour

Jacquerel committed:
Missed a spot

carlarctg commented:
This is such a high quality PR. Excellent job.

INB4 legion farm in medbay that instantly clears corpse damage.

carlarctg commented on code/modules/mob/living/basic/lavaland/legion/legion.dm:
```diff
@@ -0,0 +1,158 @@
+	ai_controller?.set_blackboard_key(BB_LEGION_RECENT_LINES, consumed.copy_recent_speech(line_chance = 80))
+	stored_mob = consumed
+	visible_message(span_warning("[src] staggers to [p_their()] feet!"))
+	if (!prob(25))
```
```suggestion
	if (prob(75))
```

i feel like thats more readable.

carlarctg commented on code/modules/mob/living/basic/lavaland/legion/legion_tumour.dm:
```diff
@@ -0,0 +1,124 @@
+		to_chat(finder, span_notice("Its tendrils seem to twitch towards the light."))
+		return
+	to_chat(finder, span_notice("Its pulsing tendrils reach all throughout the body."))
+	if(prob(10))
```
```suggestion
	if(prob(stage * 2))
```
Less likely to troll the patient if it's not about to burst 'em?

carlarctg commented on code/modules/mob/living/basic/lavaland/legion/legion_tumour.dm:
```diff
@@ -0,0 +1,124 @@
+		return FALSE
+
+	target.apply_status_effect(applied_status)
+	target.add_mood_event(MOOD_CATEGORY_LEGION_CORE, /datum/mood_event/healsbadman)
```
If I'm reading this right, it means that legion corpses have a 25% chance to have what is effectively a medium-sized regenerative core that doesn't decay. It won't zombify the body if they're not dead, which I dunno if it's intentional or not.

Cool, Ashwalkers will love this I think.

Did you make lings purge the organ with Anatomic Panacea?

Jacquerel replied:
If you bother to disembowel it out of someone I guess you get a regenerative core which doesn't decay yeah. Maybe it _should_ create a legion if used directly on a corpse which would be kind of neat.
Good catch with Panacea I forgot about it.

AtomTheProphet commented:
> Additionally, the Legion will attempt to lure your allies to Lavaland by occasionally speaking over the radio, repeating something that you said before your death.

That's creepy and I love it.

Jacquerel committed:
Update code/modules/mob/living/basic/lavaland/legion/legion_tumour.dm

Co-authored-by: carlarctg <53100513+carlarctg@users.noreply.github.com>

Jacquerel committed:
Update code/modules/mob/living/basic/lavaland/legion/legion.dm

Co-authored-by: carlarctg <53100513+carlarctg@users.noreply.github.com>

CoiledLamb commented:
incredibly pog PR, well done! how is the underlying condition you mentioned treated?

Jacquerel commented:
> incredibly pog PR, well done! how is the underlying condition you mentioned treated?

It's pretty easy, you just surgically remove the organ it implants

Jacquerel committed:
Update update paths paths
Another use for legion tumours, i hope this doesnt have any unintended consequences
Also it pulses now

Jacquerel committed:
Merge branch 'im-legion-no-im-legion' of https://github.com/Jacquerel/orbstation into im-legion-no-im-legion

Jacquerel committed:
Linter fixes

private-tristan commented:
I notice that in the gif, the legion is moving much more often. is it possible to catch up to a legion if you've hit half health slowdown? may cause problems for crusher users  being unable to reach the legion without managing to corner it.

Jacquerel commented:
> I notice that in the gif, the legion is moving much more often. is it possible to catch up to a legion if you've hit half health slowdown? may cause problems for crusher users being unable to reach the legion without managing to corner it.

Yeah you can keep up pretty easily, it stops intermittently

Jacquerel committed:
Merge branch 'master' into im-legion-no-im-legion

newfren commented:
for the record, the skulls already didn't bodyblock for the legion if you were off combat mode (but of course, you cannot rifle butt them if you're on combat mode, leading to some tedious combat mode toggling to properly fight them).

Vincent983 commented:
theorectically you could make legion farms on the station a lot easier with this, considering you can just throw a body at a legion, kill resulting legion, and take it back to station where it will reform into a legion

Jacquerel commented:
to be clear corpses left by legions don't self-ressurect, the legion tumour only progresses in stages while you are alive

Vincent983 commented:
so defib the corpse you throw at the legion
